util/expensivequery/expensivequery.go

```diff
@@ -0,0 +1,152 @@
+package expensivequery
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pingcap/log"
+	plannercore "github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/memory"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var zeroTime = time.Time{}
+
+type expensiveQueryCheckItems struct {
+	sql        string
+	sc         sessionctx.Context
+	startTime  time.Time
+	memTracker *memory.Tracker
+	plan       plannercore.Plan
+}
+
+type expensiveQueryHandler struct {
+	sync.Map
+	ticker          *time.Ticker
+	memExceedConnCh chan uint64
+}
+
+func (eqh *expensiveQueryHandler) Register(sc sessionctx.Context, sql string, startTime time.Time, plan plannercore.Plan) {
+	connID := sc.GetSessionVars().ConnectionID
+	memTracker := sc.GetSessionVars().StmtCtx.MemTracker
+	eqh.Store(connID, &expensiveQueryCheckItems{sql, sc, startTime, memTracker, plan})
+}
+
+func (eqh *expensiveQueryHandler) Unregister(connID uint64) {
+	eqh.Delete(connID)
+}
+
+// Run starts a expensive query checker goroutine at the start time of the server.
+func (eqh *expensiveQueryHandler) Run(ctx context.Context) {
+	defer eqh.ticker.Stop()
+	checkExceedTime := func(key, value interface{}) bool {
+		item := value.(*expensiveQueryCheckItems)
+		startTime := item.startTime
+		if startTime.Equal(zeroTime) {
+			return true
+		}
+		costTime := time.Since(startTime)
+		if costTime >= time.Second*time.Duration(item.sc.GetSessionVars().ExpensiveQueryTimeThreshold) {
+			// Set startTime to Zero to avoid print duplicated log.
+			item.startTime = zeroTime
+			logExpensiveQuery(ctx, item.sc, costTime, item.plan, item.sql)
+		}
+		return true
+	}
+	for {
+		select {
+		case <-eqh.ticker.C:
+			eqh.Range(checkExceedTime)
+		case connID := <-eqh.memExceedConnCh:
```

Comment:
Who send items to `memExceedConnCh`?

Reply:
The out of memory quota queries. To be finished in another PR.

Reply:
I think you can move the unrelated code to the next PR...
This is already a big one